hawkular-command-gateway-war/src/main/java/org/hawkular/cmdgw/auth/AuthenticationService.java

```diff
@@ -0,0 +1,247 @@
+
+    private final ScheduledExecutorService executorService;
+
+    /** This often all open {@link Session}s are re-authenticated */
+    private final long reauthIntervalSeconds =
+            Long.parseLong(System.getProperty(REAUTH_INTERVAL_PROP, REAUTH_INTERVAL_DEFAULT));
+
+    /** We take the lists of open {@link Session}s from here */
+    @Inject
+    private WsEndpoints wsEndpoints;
+
+    public AuthenticationService() {
+        super();
+        final int poolSize = 1;
+        this.executorService = Executors.newScheduledThreadPool(poolSize);
+
+        OkHttpClient client = new OkHttpClient();
+        client.setConnectTimeout(5, TimeUnit.SECONDS);
+        client.setReadTimeout(5, TimeUnit.SECONDS);
+        client.setWriteTimeout(5, TimeUnit.SECONDS);
+        final long keepAliveDurationNs = 500;// no reason to cache the connection open longer than 500 ms
+                                             // as our re-auth intervals are much longer
+        final int maxIdleConnections = 1;// we connect to a single URL
+        client.setConnectionPool(new ConnectionPool(maxIdleConnections, keepAliveDurationNs));
+        this.client = client;
+    }
+
+    private CloseReason reauthenticate(Session session) {
+        HttpRequestData httpRequestData =
+                (HttpRequestData) session.getUserProperties().get(HttpRequestData.class.getName());
+        if (httpRequestData == null) {
+            log.warnMissingDataForSession(HttpRequestData.class.getName(), session.getId());
+            return MISSING_HTTP_REQUESTDATA;
+        }
+        final org.hawkular.cmdgw.auth.Credentials creds =
+                (org.hawkular.cmdgw.auth.Credentials) session.getUserProperties()
+                        .get(org.hawkular.cmdgw.auth.Credentials.class.getName());
+        final String authenticationHeader = creds == null ? httpRequestData.getAuthentication()
+                : Credentials.basic(creds.getUsername(), creds.getPassword());
+        if (authenticationHeader == null) {
+            log.warnMissingDataForSession(org.hawkular.cmdgw.auth.Credentials.class.getName(), session.getId());
+            return MISSING_CREDENTIALS;
+        }
+        return reauthenticate(session.getId(), httpRequestData, authenticationHeader);
+    }
+
+    /**
+     * Performs a request against {@link AuthenticationServlet} and returns {@link #SUCCESS} in case the passed
+     * credentials are still valid or some other {@link CloseReason} if the session should be closed.
+     *
+     * @param sessionId for logging only
+     * @param httpRequestData values which were used in the {@link HttpServletRequest} that initiated the current
+     *            {@link Session}.
+     * @param authenticationHeader a HTTP header value to send to {@link AuthenticationServlet}
+     * @return {@link #SUCCESS} or some other {@link CloseReason}
+     */
+    private CloseReason reauthenticate(String sessionId, HttpRequestData httpRequestData,
```

Comment:
I'm not sure I agree with the approach here. The burden is on the client to re-authenticate itself. What I had in mind at first was to just kill the client sessions after some time and the client would just reconnect, being authenticated again by the container. This way, we get the auth for free from the container, which might or not be using basic auth. On OAuth scenarios, there would be no username or password, just bearer tokens. 

The only thing to note would be that the server should not close if there is a message pending from one side or another (ie: if the client requested a JDR report and the JDR report is still being prepared).